File: adapter-livedata/src/main/java/com/fixeads/adapter_livedata/LiveDataAdapter.java
package com.fixeads.adapter_livedata;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.LiveDataReactiveStreams;

import com.creations.livebox_common.adapter.ObservableAdapter;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Observable;

/**
 * @author Sérgio Serra.
 * sergioserra99@gmail.com
 * <p>
 * Converts an {@link Observable} to {@link LiveData} instance using {@link LiveDataReactiveStreams}
 */
public class LiveDataAdapter<T> implements ObservableAdapter<T, LiveData<T>> {

    @Override
    public LiveData<T> adapt(Observable<T> observable) {
        return LiveDataReactiveStreams.fromPublisher(observable.toFlowable(BackpressureStrategy.BUFFER));
    }

}

File: adapter-livedata/src/main/java/com/fixeads/adapter_livedata/LiveDataAdapter.kt
package com.fixeads.adapter_livedata

import android.annotation.SuppressLint
import androidx.lifecycle.LiveData
import androidx.lifecycle.LiveDataReactiveStreams
import com.creations.livebox_common.adapter.ObservableAdapter
import com.creations.runtime.state.State
import com.creations.runtime.state.error
import com.creations.runtime.state.loading
import com.creations.runtime.state.success
import io.reactivex.BackpressureStrategy
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.Subject
import java.util.concurrent.atomic.AtomicBoolean


/**
 * @author Sérgio Serra.
 * Converts an [Observable] to [LiveData] instance using [LiveDataReactiveStreams]
 */
class LiveDataAdapter<T> : ObservableAdapter<T, LiveData<T>> {
    override fun adapt(observable: Observable<T>): LiveData<T> =
            LiveDataReactiveStreams.fromPublisher(observable.toFlowable(BackpressureStrategy.BUFFER))
}

class StateAdapter<T> : ObservableAdapter<T, Observable<State<T>>> {

    val isSubscribed = AtomicBoolean(false)

    @SuppressLint("CheckResult")
    override fun adapt(observable: Observable<T>): Observable<State<T>> {
        val subject: Subject<State<T>> = BehaviorSubject.create<State<T>>().toSerialized()
        val obs = subject.doOnSubscribe { _ ->
            if (isSubscribed.compareAndSet(false, true)) {
                observable
                        .subscribeOn(Schedulers.io())
                        .doOnSubscribe { subject.onNext(loading()) }
                        .subscribe({
                            subject.onNext(success(it))
                        }, {
                            subject.onNext(error())
                        })
            }
        }
        return obs
    }
}
